Buffer logs before post (version 0.2.0)

brtkwr opened the pull request.

Commit:
Buffer logs before post

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -34,7 +36,7 @@
     """Client for interacting with the monasca-log-api."""
     
-    def __init__(self):
+    def __init__(self, interval=10):
```
Please could we make this a config variable?

brtkwr replied:
Done

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +48,22 @@ def __init__(self):
         self._url = cfg.CONF.api.url
+        self.buffer = {}
+        self.interval = interval
+        self.start_time = time.time()
```
Would we be able to flush the buffer if it exceeds some maximum size as set in the config file?

If it's simpler, we could just define a maximum number of messages which can be held before a flush.

brtkwr replied:
I've added a max_buffer_size variable.

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +48,22 @@ def __init__(self):
+        for key, value in jsonutils.loads(data).items():
+            self.buffer.setdefault(key, []).extend(value)
+        elapsed_time = time.time() - self.start_time
+        if elapsed_time > self.interval:
```
I think this works fine if you have a steady inflow of logs, but what if you have a single critical log file that comes in, and then nothing for a few hours. Won't that critical log file sit in the buffer until another log message comes in? Do we need to periodically flush this, so that we can guarantee that logs won't sit in the buffer for much longer than the interval?

dougszumski replied:
Any thoughts on this?

brtkwr replied:
To answer your question, I have put in a timeout mechanism using a non configurable `proc_interval` of 1 second in the `stdin_by_lines` function which yields None and allows items in the queue to be flushed if the acceptable interval has elapsed before it is called again.

dougszumski reviewed (commented):
Thanks Bharat! Some questions  inline.

Commit:
Read max_interval, min_interval and max_buffer_size from config file

Commit:
Improve logic for processing queues when they are filling up too quickly

Commit:
Configurable option to print additional log

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +61,46 @@ def __init__(self):
+        self._max_buffer_size = cfg.CONF.api.max_buffer_size
+        self.log_count, self.log_buffer, self.start_time = 0, {}, time.time()
+        # Allow time for buffer to build before polling rsyslog
+        time.sleep(self._poll_interval)
```
If this is being instantiated once by rsyslog, how does this work, or is that not the case?

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +61,46 @@ def __init__(self):
+        # Allow time for buffer to build before polling rsyslog
+        time.sleep(self._poll_interval)
 
     def post_logs(self, data):
```
How is this called by rsyslog? Is it only when a log message is received?

brtkwr replied:
Looks like inputs are piped into the main script. The daemon creates a client connection when it starts and posts json blobs to API endpoint until it is stopped or there is a terminal failure at which point it restarts.

dougszumski reviewed (commented):
Thanks for updating

Commit:
Hand over control to client

This is in the event when there is no input logs coming in but there are
messages in the buffer that are waiting to be flushed.

Commit:
Catch connection failure and keep retrying until success

Commit:
Hard code a proc_interval of 1 second

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +59,58 @@ def __init__(self):
+        # Allow time for buffer to build before polling rsyslog
+        time.sleep(proc_interval)
+
+        # Reset the variables
```
Do you mean initialise?

brtkwr replied:
Ah yes

Commit:
Catch all post related errors and flush to stderr

Commit:
Break code into functional units to make things more readable

Commit:
Indicate when about to take a nap in verbose mode

Commit:
Fix for edge case when max_buffer_size=1

Commit:
Remove proc_interval altogether and use min_interval alone to poll stdin

dougszumski commented:
I took this for a spin and it works nicely out of the box. I generated a bunch of test logs  with a small script:

```
#!/bin/bash         
while :             
do                  
  logger test `date`
done                
```

Output:

```
==> /var/log/monasca-rsyslog.stdout <==
log_count       1       elapsed_time    10.1014630795   waited_too_long True    buffer_too_long False
log_count       2       elapsed_time    10.2438120842   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.6839089394   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.9635839462   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.4149410725   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.188988924    waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.2325959206   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.9328980446   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.5787949562   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.8959350586   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.83345294     waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.6202747822   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.4750277996   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.4344260693   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.322947979    waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.3067650795   waited_too_long True    buffer_too_long False
log_count       5       elapsed_time    10.5181250572   waited_too_long True    buffer_too_long False
log_count       15      elapsed_time    10.3680038452   waited_too_long True    buffer_too_long False
log_count       2       elapsed_time    10.2452399731   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.181511879    waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.8339920044   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.5681810379   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.952755928    waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.9651069641   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.7883820534   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    15.7170240879   waited_too_long True    buffer_too_long False
log_count       6       elapsed_time    17.8246259689   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    13.2439620495   waited_too_long True    buffer_too_long False
log_count       4       elapsed_time    11.1647870541   waited_too_long True    buffer_too_long False
log_count       101     elapsed_time    0.279726982117  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.264189958572  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.254410028458  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.248872995377  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.282093048096  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.259469032288  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.246901988983  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.257671833038  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.250741958618  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.239701032639  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.25283408165   waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.250095844269  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.265928030014  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.255164146423  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.256454944611  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.25420498848   waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.251795053482  waited_too_long False   buffer_too_long True
log_count       101     elapsed_time    0.246290922165  waited_too_long False   buffer_too_long True
log_count       94      elapsed_time    15.1953141689   waited_too_long True    buffer_too_long False
log_count       3       elapsed_time    16.0089688301   waited_too_long True    buffer_too_long False
```

The logs appear in Kibana fine.

brtkwr commented:
Good to hear :)

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +59,67 @@ def __init__(self):
+            if (waited_too_long or buffer_too_long) and log_count > 0:
+                # Print a summary of states
+                info_keys = ["log_count", "elapsed_time", "waited_too_long", "buffer_too_long"]
+                print('\t'.join(["{}\t{}".format(k, locals()[k]) for k in info_keys]))
```
What do you think about only printing this if verbose mode is on so we don't need to worry about slowly generating a massive log file?

brtkwr replied:
Agree, I've changed verbose from boolean to verbosity which is an integer to control different levels of verbosity.

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -26,13 +29,23 @@
-               help='Specify URL for the logging API')
+               help='Specify URL for the logging API'),
+    cfg.IntOpt('max_batch_size',
+               default=100,
```
Seems like a reasonable choice. We should generally keep posts under 1MB with this.

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +59,67 @@ def __init__(self):
+                sys.stdout.flush()
+                sys.stderr.write(e)
+                sys.stderr.flush()
+                time.sleep(1)
```
Good idea putting this in - if the API goes down this will prevent vast amounts of `connection failed` log spam.

dougszumski commented on monasca_rsyslog/main.py:
```diff
@@ -23,30 +23,31 @@
-        line = sys.stdin.readline()
-        if line:
-            yield line
+        buffer_is_not_empty, _, _ = select([sys.stdin], [], [], poll_interval)
```
Interesting approach!

brtkwr replied:
Thanks, seems to work surprisingly well!

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -26,13 +29,23 @@
+    cfg.IntOpt('min_poll_delay',
+               default=10,
+               help='Specify minimum time to wait between triggering a post'),
+    cfg.BoolOpt('verbose',
```
nit: Maybe this should live outside the API section in the conf file

brtkwr replied:
We can deal with this when we infer the api URL via keystone.

dougszumski reviewed (commented):
Nice job, it works well in testing.

Please could we update the example config file?

https://github.com/stackhpc/monasca-rsyslog/blob/master/etc/monasca/monasca-rsyslog.conf

Commit:
Only expose handle_logs function publicly

Commit:
Introduce verbosity levels rather than verbose boolean to have different level of logs

Commit:
Update example config

Commit:
Move stdin_by_line function to the Client class

dougszumski commented on monasca_rsyslog/client.py:
```diff
@@ -46,13 +60,83 @@ def __init__(self):
+            buffer_too_long = log_count > self._max_batch_size
+            if (waited_too_long or buffer_too_long) and log_count > 0:
+                # Print a summary of states
+                if self._verbosity > 0:
```
nit: This stanza could sit in a function since we use it multiple times (or use an external logging library), but not to worry.

brtkwr replied:
Good idea, done!

Commit:
Move summary printer to its own function
